### PR TITLE
fix: localmode subprocess parent process not sending SIGTERM to child

### PR DESCRIPTION
**Pull request by ranzvi**
*Issue #, if available:*

*Description of changes:*

Due to an unknown issue, on some machines executing `predictor.delete_endpoint()` simply hangs.
This is because the `SIGTERM` sent to the subprocess doesn't propagate to the child process.

image

E.g pid `7521` doesn't propagate `SIGTERM` to child `7522`.

This means that `self.container.join()` is waiting for a container which hasn't received a signal to stop execution and hangs indefinitely.

By sending `SIGTERM` to all child processes of the `subprocess.Popen` process, we can guarantee that the container will be Gracefully stopped...

Testing done:

Executed all unit and integration tests
Verified behaviour remains the same on machines without the bug
Verified bug fixed on faulty machines by executing multiple endpoints sequentially
*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by ranzvi** [comment]
@ahsan-z-khan This PR doesn't contain and new dependencies

**Comment by gilinachum** [comment]
@shreyapandit Hi! can you have a look if the failing tests are relevant to the PR? And assign someone to review it?

**Review by BasilBeirouti — DISMISSED** [review]

**Comment by shreyapandit** [comment]
Hi @ranzvi, thanks for your contribution!

Please follow the one time setup for enabling git hooks, and this should allow you to reproduce the errors, and fix the tests cc @ranzvi @gilinachum. The unit tests are failing on linting issues. 
Here is the setup can you do. Please change brew install to whatever is compatible with your OS.

```
# Install using brew:
brew install pyenv pyenv-virtualenv pyenv-virtualenvwrapper

# Make sure to add the pyenv and virtualenv init functions to your corresponding shell init (.bashrc, .zshrc, etc):
eval "*$(*pyenv init --path*)*"      # old usage was: $ eval "$(pyenv init -)"
eval "*$(*pyenv virtualenv-init -*)*"

# Start a new shell once you do that to pick up your changes.
# Install following Python versions:
pyenv install 3.8.8
pyenv install 3.7.10
pyenv install 3.6.13

# Set them as global versions:
pyenv global 3.8.8
pyenv global 3.7.10
pyenv global 3.6.13

# Verify they show up when you do:
pyenv versions

# Restart your shell and run the command again to verify that it persists across shell sessions.

# Install tox to run our tests:
pip install tox

# Run the tests by running:
# tox

# Set all versions as global:
pyenv global $(pyenv versions --bare)
pyenv rehash

# Setup GitHub Pre Push hooks
# To enable all git hooks in the .githooks directory.
# Run these commands in the GitHub repository directory:

find .git/hooks -type l -exec rm {} \;
find .githooks -type f -exec ln -sf ../../{} .git/hooks/ \;```

**Review by shreyapandit — CHANGES_REQUESTED** [review]
As outlined in comments

File: src/sagemaker/local/image.py [hunk 1]
```diff
@@ -841,6 +840,8 @@ def run(self):
 
     def down(self):
         """Placeholder docstring"""
+        if os.name != "nt":
```
[the remark this answers is not in the document]
Comment (ranzvi) [reply src/sagemaker/local/image.py:843]: It's checking if the os is a windows machine, since I'm using unix commands to kill the processes

**Comment by ranzvi** [comment]
@shreyapandit Thanks for the insights. I fixed the linting issues and executed the unittest locally and they pass.

Do you have any idea why the rest of the ci is still failing?

**Comment by navinsoni** [comment]
@ranzvi I am restarting your tests

**Review by shreyapandit — APPROVED** [review]
LGTM! Thanks for your contribution

**Review by jeniyat — APPROVED** [review]
looks good to me.

**Comment by ranzvi** [comment]
@wcarpenter1-godaddy This did fix our issue, we had a problem with the `self.process.terminate()` which is why I added `kill_child_processes`. Could you investigate where this error is originating and why?

**Comment by wcarpenter1-godaddy** [comment]
@ranzvi what os are you on?

**Comment by ranzvi** [comment]
@wcarpenter1-godaddy MacOS, but this works on `ubuntu` as well

**Pull request merged**